cluster-autoscaler/estimator/binpacking_estimator.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package estimator
+
+import (
+	"sort"
+
+	"k8s.io/contrib/cluster-autoscaler/simulator"
+	kube_api "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+)
+
+// podInfo contains Pod and score that corresponds to how important it is to handle the pod first.
+type podInfo struct {
+	score float64
+	pod   *kube_api.Pod
+}
+
+type byScoreDesc []*podInfo
+
+func (a byScoreDesc) Len() int           { return len(a) }
+func (a byScoreDesc) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byScoreDesc) Less(i, j int) bool { return a[i].score > a[j].score }
+
+// BinpackingNodeEstimator estimates the number of needed nodes to handle the given amount of pods.
+type BinpackingNodeEstimator struct {
+	predicateChecker *simulator.PredicateChecker
+}
+
+// NewBasicNodeEstimator builds BasicNodeEstimator.
+func NewBinpackingNodeEstimator(predicateChecker *simulator.PredicateChecker) *BinpackingNodeEstimator {
+	return &BinpackingNodeEstimator{
+		predicateChecker: predicateChecker,
+	}
+}
+
+// Estimate implements First Fit Decreasing bin-packing approximation algorithm.
+// See https://en.wikipedia.org/wiki/Bin_packing_problem for more details.
+// While it is a multi-dimensional bin packing (cpu, mem, ports) in most cases the main dimension
+// will be cpu thus the estimated overprovisioning of 11/9 * optimal + 6/9 should be
+// still be maintained.
+// It is assumed that all pods from the given list can fit to nodeTemplate.
+func (estimator *BinpackingNodeEstimator) Estimate(pods []*kube_api.Pod, nodeTemplate *schedulercache.NodeInfo) int {
+	podInfos := make([]*podInfo, 0, len(pods))
+
+	// Calculate score for all pods.
+	// Score is defined as cpu_sum/node_capacity + mem_sum/node_capacity. With this pods that
```

Comment:
Nitpick:
How about rephrasing second sentence to
"Pods that have bigger requirements should be processed first, thus have higher scores."